torch/_dynamo/codegen.py

```diff
@@ -158,9 +159,27 @@ def __call__(self, value, allow_cache=True):
         if isinstance(value, Source):
             # If the source needs to be overridden, use the new one.
             source = self.overridden_sources.get(value, value)
-            self.call_reconstruct(source)
-            # We don't support dup_top optimization for source yet.
-            self.clear_tos()
+            if allow_cache:
+                if self.top_of_stack is value:
+                    self._output.append(create_dup_top())
+                    return
+
+                if self.tempvars.get(source) is not None:
+                    self._output.append(self.create_load(self.tempvars[source]))
+                    self.top_of_stack = source
+                    return
+
+            try:
+                self.call_reconstruct(source)
+            except NotImplementedError:
+                unimplemented(f"reconstruct: {source}")
+
+            if allow_cache and source in self.tempvars:
+                self._output.append(create_dup_top())
+                self.add_cache(source)
+
+            self.top_of_stack = source
```

Comment:
This `allow_cache` flag really hurts brain lol. I believe it'll always be True here, and maybe it's worth keeping/asserting that (documented) semantics, just to contain the subtlety to VT only..